Explicitly test all arguments in the constructor to fix problems when…

jeff-kilbride opened the pull request:
… two arguments are passed.

Hi Matteo,

Thank you for contributing the split2 module! I am using it to parse large newline delimited json files and insert them into a MongoDB database.

I have made some changes to the split() function to explicitly test for the arguments passed. I would like to be able to pass two arguments to split() and have it work correctly. For example, with the original code, the following will not work:

```javascript
var fs = require('fs');
var split2 = require('split2');
var reader = fs.createReadStream('./in.json');
var writer = fs.createWriteStream('./out.json');

reader.pipe(split2(JSON.parse, { highWaterMark: 65535 })).pipe(writer);
```

In this example, I want to use the default matcher and supply a mapper and options. However, with the original code the options are lost. Also, this will not work:

```javascript
reader.pipe(split2("\n", { highWaterMark: 65535 })).pipe(writer);
```

In the above, I want to supply a matcher and options but use the default mapper. Doing this causes an error:

```javascript
    push(this, this.mapper((this._last + list.shift())))
                    ^
TypeError: this.mapper is not a function
    at DestroyableTransform.transform [as _transform] (/home/jeff/WebstormProjects/Streams/node_modules/split2/index.js:27:21)
    at DestroyableTransform.Transform._read (/home/jeff/WebstormProjects/Streams/node_modules/split2/node_modules/through2/node_modules/readable-stream/lib/_stream_transform.js:172:10)
    at DestroyableTransform.Transform._write (/home/jeff/WebstormProjects/Streams/node_modules/split2/node_modules/through2/node_modules/readable-stream/lib/_stream_transform.js:160:12)
    at doWrite (/home/jeff/WebstormProjects/Streams/node_modules/split2/node_modules/through2/node_modules/readable-stream/lib/_stream_writable.js:333:12)
    at writeOrBuffer (/home/jeff/WebstormProjects/Streams/node_modules/split2/node_modules/through2/node_modules/readable-stream/lib/_stream_writable.js:319:5)
    at DestroyableTransform.Writable.write (/home/jeff/WebstormProjects/Streams/node_modules/split2/node_modules/through2/node_modules/readable-stream/lib/_stream_writable.js:246:11)
    at ReadStream.ondata (_stream_readable.js:528:20)
    at emitOne (events.js:77:13)
    at ReadStream.emit (events.js:169:7)
    at readableAddChunk (_stream_readable.js:146:16)
```

I can get around this by supplying all three arguments in both cases:

```javascript
reader.pipe(split2(/\r?\n/, JSON.parse, { highWaterMark: 65535 })).pipe(writer);
```
OR
```javascript
reader.pipe(split2("\n", null, { highWaterMark: 65535 })).pipe(writer);
```

However, I thought it would be nice for the split() function to recognize when two arguments are passed and do this automatically. Let me know what you think.

Thanks again!
--jeff

yoshuawuyts commented:
Regardless of functionality, there seems to be a lot of noise / stylistic changes in this PR. It's good form to keep the style intact when submitting a patch regardless of personal opinion, if anything it if makes it easier for maintainers to respond / judge the quality of your patch.

jeff-kilbride commented:
Yeah, sorry. This is my first PR in github. My environment is setup with default tab and indentation at 4, whereas the original was at 2. I thought I had converted it all back to the original format, but didn't catch the extra spaces in the license and one of the declarations until after I had created the PR. The only functional changes are in split(). The whitespace changes were unintentional and created by my IDE's auto-formatting. My bad.

I can resubmit with just my changes to the split() function, if necessary. Let me know.

yoshuawuyts commented:
(': all good, glad you're taking it positively. I think if you modify the values and force push override the branch you should be all good. Also: this project doesn't do semicolons, hah. Cheers!

jeff-kilbride commented:
Ok, I pushed a new commit to my repo and it looks like it was automatically picked up in the PR. All the whitespace changes are gone and I removed my semicolons -- but I left the one on line 96 that was in the original...    :^)

mcollina commented:
would you mind adding some unit tests for this? I would like to avoid possible regressions.

jeff-kilbride commented:
Sure. I will try adding some unit tests to your test.js file, unless you would prefer a different method.

mcollina commented:
Yes, that would be perfect
Il giorno mer 23 dic 2015 alle 18:06 Jeff Kilbride <notifications@github.com>
ha scritto:

> Sure. I will try adding some unit tests to your test.js file, unless you
> would prefer a different method.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mcollina/split2/pull/5#issuecomment-166945580>.
>

jeff-kilbride commented:
I have added three tests to test.js:

support custom matcher and mapper
support custom matcher and options
support mapper and options

This should cover all two argument variations for the constructor.

Also, I had initially copied your "support an option argument" test for my options tests, but noticed that the highWaterMark option was not being set correctly. On line 75, you had "highWatermark" (notice the small "m"). So, I added explicit testing for the matcher, mapper, and options values being set and corrected the typo on line 75.

mcollina commented:
This is very well done, thank you so much!!

mcollina commented:
Thanks @yoshuawuyts for reviewing this :)

yoshuawuyts commented:
Glad to see it landed, looks very neat! Happy to have been of help 🐙

jeff-kilbride commented:
You're very welcome! Glad to be of service.
